cub/test/catch2_test_device_partition_flagged.cu

```diff
@@ -0,0 +1,351 @@
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = 0;
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  thrust::device_vector<char> flags(num_items);
+
+  // Needs to be device accessible
+  thrust::device_vector<int> num_selected_out(1, 0);
+  int    *d_num_selected_out = thrust::raw_pointer_cast(num_selected_out.data());
+
+  partition_flagged(in.begin(),
+                    flags.begin(),
+                    out.begin(),
+                    d_num_selected_out,
+                    num_items);
+
+  REQUIRE(num_selected_out[0] == 0);
+}
+
+CUB_TEST("DevicePartition::Flagged handles all matched", "[device][partition_flagged]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  thrust::device_vector<char> flags(num_items);
+  thrust::fill(flags.begin(), flags.end(), 1);
+
+  // Needs to be device accessible
+  thrust::device_vector<int> num_selected_out(1, 0);
+  int    *d_num_selected_out = thrust::raw_pointer_cast(num_selected_out.data());
+
+  partition_flagged(in.begin(),
+                    flags.begin(),
+                    out.begin(),
+                    d_num_selected_out,
+                    num_items);
+
+  REQUIRE(num_selected_out[0] == num_items);
+  REQUIRE(out == in);
+}
+
+CUB_TEST("DevicePartition::Flagged handles no matched", "[device][partition_flagged]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  thrust::device_vector<char> flags(num_items);
+  thrust::fill(flags.begin(), flags.end(), 0);
+
+  // Needs to be device accessible
+  thrust::device_vector<int> num_selected_out(1, 0);
+  int    *d_num_selected_out = thrust::raw_pointer_cast(num_selected_out.data());
+
+  partition_flagged(in.begin(),
+                    flags.begin(),
+                    out.begin(),
+                    d_num_selected_out,
+                    num_items);
+
+  // The false partition is in reverse order
+  thrust::reverse(out.begin(), out.end());
+
+  REQUIRE(num_selected_out[0] == 0);
+  REQUIRE(out == in);
+}
+
+CUB_TEST("DevicePartition::Flagged does not change input", "[device][partition_flagged]", types)
+{
+  using type = typename c2h::get<0, TestType>;
+
+  const int num_items = GENERATE_COPY(take(2, random(1, 1000000)));
+  thrust::device_vector<type> in(num_items);
+  thrust::device_vector<type> out(num_items);
+  c2h::gen(CUB_SEED(2), in);
+
+  const int num_selected                    = get_num_selected(num_items);
+  const thrust::device_vector<char> flags   = get_flags<char>(num_items, num_selected);
```

Comment:
```suggestion
  thrust::device_vector<char> flags(num_items);
  c2h::gen(CUB_SEED(1), flags);
```

Reply:
I am not comfortable with that change, as the chance that flags are actually zero is rather limited. 

We already test flags that are only convertible to bool, so I feel this would be a regression

Reply:
Agreed. Could you adopt the following [code](https://github.com/NVIDIA/cccl/blob/99a3a0d4c906e236877de7b24c4f2b4195f2354e/cub/benchmarks/nvbench_helper/nvbench_helper.cu#L360C8-L382) to our generators so that we can generate bool? Alternatively, you might consider providing min/max range to gen, so that the generation happens in range `[0, 1]`.

Reply:
We talked about this internally and decided that for now we will move forward with `c2h::gen(CUB_SEED(1), flags, 0, 1);`

This will give us a 50% probability of elements being moved